### PR TITLE
feat: improve performance BTC reward distribution

### PR DESCRIPTION
**Pull request by RafilxTenfen**
- Add new structures to track BTC delegation rewards and lazily distribute
- Add new event `BTCDelegationStatus_EXPIRED` for BTC delegations that indicates a BTC delegation is no longer contributing to security 
- Important files diff for review
  -  `proto/babylon/incentive/rewards.proto`
  -  `x/finality/keeper/power_dist_change.go`
  - `x/incentive/keeper/btc_staking_gauge.go`
  - `x/incentive/keeper/grpc_query.go`
  - `x/incentive/keeper/msg_server.go`
  - `x/incentive/keeper/reward_tracker.go`
  - `x/incentive/keeper/reward_tracker_store.go`
  - `x/btcstaking/keeper/btc_delegations.go`
  - `x/btcstaking/keeper/msg_server.go`
  - `x/btcstaking/types/btc_delegation.go`

## Benchmarks

Command to generate pprof `go tool pprof -http=:8080 http://localhost:6060/debug/pprof/profile\?seconds\=300`

- Main branch that iterates over delegations to distribute rewards with 163k BTC delegations, took 15.52s (13.1%) of the time during the 300seconds interval used to generate the pprof 

image

- Current branch `rafilx/improve-performance-btc-dist` with 100k BTC delegations, took 6.73s (1.8%) of the time during the 300seconds interval

image

Note that the percentage of the time used that is ~7x lower should be considered instead of the time, the benchmark from `main` was running in the benchmark server 48 processors and 256GB of RAM while the one in the current branch I run locally with 24 processors and 96GB of RAM.... So the percentage of the time usage instead of seconds should be the one in comparison

> Thanks @Lazar955  for help with benchmark
> The function used to show the difference is the `TallyBlocks` because the `RewardBTCStaking` does not appear in the pprof of the current branch demonstrating that the time spent there is negligible to the overall period

### CODE REVIEW

**Review by KonradStaniec — COMMENTED** [review]
Some initial comments, I am still parsing how it should supposed to work tbh.

It would be great to have some e2e test with pre-defined values i.e setup some scenario with  lets ay 2 fps, each fp having 2 delegations with pre-defined values, known inflation, and split between fps and validators and calculate whether caluclated rewards for stakers makes sense.

File: x/finality/keeper/msg_server_test.go
Comment (RafilxTenfen) [reply x/finality/keeper/msg_server_test.go:490]: This is the integration test, planning to finalize today

File: x/finality/keeper/power_dist_change.go
Comment (RafilxTenfen) [reply x/finality/keeper/power_dist_change.go:356]: nop, it is not
If we call twice, it will reduce the voting power twice

File: x/finality/keeper/power_dist_change.go
Comment (RafilxTenfen) [reply x/finality/keeper/power_dist_change.go:266]: this is handle by the list of actives FPs, that will not receive rewards for the block that they were jailed 
https://github.com/babylonlabs-io/pm/pull/141#discussion_r1879170625

https://github.com/babylonlabs-io/babylon/blob/dd4009df6b7fc6a9a45f56481485f5980f6d6873/x/finality/types/power_table.go#L68

This is the part where it have as active the top N FPs  and these are the ones that are going to receive rewards

File: x/finality/keeper/power_dist_change.go
Comment (RafilxTenfen) [reply x/finality/keeper/power_dist_change.go:250]: not sure on this one 
@SebastianElvis ?

File: testutil/incentives-helper/keeper.go
Comment (RafilxTenfen) [reply testutil/incentives-helper/keeper.go:27]: Removing it

File: x/finality/keeper/power_dist_change.go
Comment (RafilxTenfen) [reply x/finality/keeper/power_dist_change.go:356]: > Asking, as there is slight edge case in unbonding in which we can process two unboding events for one delegation.

Can you give more details about this one?

Comment (RafilxTenfen) [reply x/finality/keeper/power_dist_change.go:356]: okay, this would break the reward tracker 
checking what could I do here

Comment (RafilxTenfen) [reply x/finality/keeper/power_dist_change.go:361]: The ubonding events for the reward tracker was modified to be processed only to the cases where the btc delegation transactions are already inside `FinalityProviderDistInfo`

https://github.com/babylonlabs-io/babylon/blob/9ee3ba50f54f8f09257aef4a8e925da010c814a4/x/finality/keeper/power_dist_change.go#L295

@gitferry do you see any problem with this?

**Comment by RafilxTenfen** [comment]
> It would be great to have some e2e test with pre-defined values i.e setup some scenario with lets ay 2 fps, each fp having 2 delegations with pre-defined values, known inflation, and split between fps and validators and calculate whether caluclated rewards for stakers makes sense.

I have added [an integration tests with rewards](https://github.com/babylonlabs-io/babylon/blob/01479f330ea98f1813eda655a4a56c5232680874/x/finality/keeper/msg_server_test.go#L496), and there is also a fuzz test that verifies the proportion distributed [FuzzRewardBTCStaking](https://github.com/babylonlabs-io/babylon/blob/aafda528440b6de3b4da254420813e57e7a261e4/x/incentive/keeper/btc_staking_gauge_test.go#L16) 

but I will write an e2e test tomorrow with 2 active fps and 2 BTC delegations, it will be a bit difficult to estimate the rewards, but I think the goal would be to see the difference between the delegations amounts

File: app/app.go
Comment (SebastianElvis) [inline app/app.go:145]: Why is this made public?

File: proto/babylon/incentive/rewards.proto
Comment (SebastianElvis) [inline proto/babylon/incentive/rewards.proto:15]: Does this sentence mean `cumulative_rewards_per_sat` = cumulative rewards during the given period / amount of Sats delegated to this FP?

File: testutil/coins/margin.go
Comment (SebastianElvis) [inline testutil/coins/margin.go:11]: Can this be moved to `margin_test.go` as this is a test util func?

File: testutil/incentives-helper/keeper.go
Comment (SebastianElvis) [inline testutil/incentives-helper/keeper.go:1]: This file seeems to have large overlap with `testutil/btcstaking-helper` or `testutil/helper`. Any reason why we cannot use those?

File: x/incentive/keeper/reward_tracker.go
Comment (SebastianElvis) [inline x/incentive/keeper/reward_tracker.go:59]: Why do we need to enforce withdrawal and remove the BTC delegation reward trackers? Upon slashing this FP will lose all voting power and won't get any reward anyway. It seems secure to keep the rewards in gauge as is

File: x/incentive/keeper/reward_tracker.go
Comment (SebastianElvis) [inline x/incentive/keeper/reward_tracker.go:73]: Yeah tbh I'm not sure if this is necessary, and this iterating process might be some perf hotspot (It iterates over all BTC delegations under a slashed FP).

File: x/finality/keeper/power_dist_change.go
Comment (SebastianElvis) [reply x/finality/keeper/power_dist_change.go:250]: No `BtcDels` field is no longer necessary. It is used for tracking the share of every BTC delegation under a given FP. Now this shall be replaced by `storeBTCDelegationRewardsTracker` store

File: x/incentive/keeper/reward_tracker.go
Comment (SebastianElvis) [inline x/incentive/keeper/reward_tracker.go:59]: Do we need to do anything upon a slashed FP? 

- After a FP is slashed BTC delegators will unbond and withdraw anyway, and this will trigger all those lazy executions. So enforce withdrawal does not seem necessary.
- Enforcing withdrawal might cause some edge case, e.g., a FP is slashed which triggers all withdrawal. Then if a BTC delegator wants to withdraw will the code panick since some keys in KV store are removed?

File: x/incentive/keeper/grpc_query.go
Comment (SebastianElvis) [inline x/incentive/keeper/grpc_query.go:32]: RPC should not trigger state update at all. Otherwise it's possibel that someone queries this API and the queried node's state root is updated, then in the next block this node won't have a matching state root with any other node.

File: x/finality/keeper/power_dist_change.go
Comment (SebastianElvis) [inline x/finality/keeper/power_dist_change.go:286]: Would recommend we remove any usage of `BtcDels` under `FinalityProviderDistInfo`. The original purpose of `BtcDels` is to track the share of every BTC delegator under the FP. Now this is replaced by KV stores under `x/incentive` and no need to do this anymore

File: x/finality/keeper/power_dist_change.go
Comment (SebastianElvis) [inline x/finality/keeper/power_dist_change.go:353]: Can we do this inside the `for _, event := range events` loop? This loop was supposed to be the "event bus" that dispatches notifications to relevant modules, including `x/incentive`

**Review by SebastianElvis — COMMENTED** [review]
That's a comprehensive PR :+1: Still parsing this. Some initial comments

Meanwhile, is it possible to consolidate some of the abstractions, e.g., in `reward_tracker.go` and `reward_tracker_store.go`? The current implementation is a bit hard to follow and some abstractions might not be necessary. See comments

File: app/app.go
Comment (RafilxTenfen) [reply app/app.go:145]: just to import in a test, I can duplicate the code to avoid exposure

File: proto/babylon/incentive/rewards.proto
Comment (RafilxTenfen) [reply proto/babylon/incentive/rewards.proto:15]: well it is the previous FinalityProviderHistoricalRewards.CumulativeRewardsPerSat + the rewards of the given period 

That was the reason I called it cumulative, in the cosmos-sdk they had the equivalent structure that was storing the amount of rewards per share of voting power and they called [CumulativeRewardRatio](https://github.com/cosmos/cosmos-sdk/blob/d9c53bfefc1e75a3c6b09065ea8b3a836cda0d18/x/distribution/types/distribution.pb.go#L97)

File: testutil/coins/margin.go
Comment (RafilxTenfen) [reply testutil/coins/margin.go:11]: if moved to a test, it can't be imported by another package

File: testutil/incentives-helper/keeper.go
Comment (RafilxTenfen) [reply testutil/incentives-helper/keeper.go:1]: I did use the btcstaking-helper as a base helper inside the incentives structure

File: proto/babylon/incentive/rewards.proto
Comment (RafilxTenfen) [reply proto/babylon/incentive/rewards.proto:64]: Agree :+1:  this could be done, I just think this PR is already to big to introduce new changes

File: x/incentive/keeper/grpc_query.go
Comment (RafilxTenfen) [reply x/incentive/keeper/grpc_query.go:32]: in the query the database does not commit the modified changes to the state 
The same is done by the cosmos-sdk [DelegationRewards](https://github.com/cosmos/cosmos-sdk/blob/46be66ed4205bf109d93704dd14c0822d048972e/x/distribution/keeper/grpc_query.go#L244-L245) query

File: x/finality/keeper/power_dist_change.go
Comment (RafilxTenfen) [reply x/finality/keeper/power_dist_change.go:353]: we could if the events were sorted and never received more than once 

ref https://github.com/babylonlabs-io/babylon/pull/306#discussion_r1886719715

One BTC delegation can generate 2 unbonding events

File: x/incentive/keeper/reward_tracker.go
Comment (RafilxTenfen) [reply x/incentive/keeper/reward_tracker.go:59]: This is not strictly necessary, but the FP will not receive rewards anymore under that BTC pk, so we will basically stay with all of those structures inside the database, my idea here was to use everything at once and then prune it 

> Enforcing withdrawal might cause some edge case, e.g., a FP is slashed which triggers all withdrawal. Then if a BTC delegator wants to withdraw will the code panick since some keys in KV store are removed?

If the BTC delegator wants to withdraw after the FP is slashed, it will just not exists that FP,DEL record anymore and will not appear in the iterator, it should not panic

File: x/finality/keeper/power_dist_change.go
Comment (gitferry) [reply x/finality/keeper/power_dist_change.go:361]: I think this pr works because the unbonded delegation will not appear in the cache so that unbonding events will processed twice by the incentive module

File: x/finality/keeper/power_dist_change.go
Comment (gitferry) [inline x/finality/keeper/power_dist_change.go:360]: it seems that `cacheFpByBtcPkHex` is sorted twice in each processor. Can we pass a sorted slice here?

**Review by gitferry — COMMENTED** [review]
Great work! Some inital comments:

File: app/app.go
Comment (RafilxTenfen) [reply app/app.go]: rollback to private maccPerms

File: x/incentive/keeper/reward_tracker.go
Comment (RafilxTenfen) [reply x/incentive/keeper/reward_tracker.go:150]: We could, but it is very helpful to have a separate function that only calculates the rewards instead of calculating and sending them to the gauge

File: proto/babylon/incentive/rewards.proto
Comment (RafilxTenfen) [reply proto/babylon/incentive/rewards.proto:15]: updated to `// The cumulative rewards of that finality provider per sat until that period`

File: x/btcstaking/types/btc_delegation.go
Comment (RafilxTenfen) [reply x/btcstaking/types/btc_delegation.go:62]: staker

File: x/finality/keeper/power_dist_change.go
Comment (RafilxTenfen) [reply x/finality/keeper/power_dist_change.go:361]: it shouldn't be processed twice, only once

Comment (RafilxTenfen) [reply x/finality/keeper/power_dist_change.go:361]: I will remove the btcDelegations inside the `FinalityProviderDistInfo` and add a new key inside the `x/finality` that stores the tx hash of already processed unbonded events so that expired with early unbending can be dropped when received by the second time

What do you guys think? @babylonlabs-io/core-dev

File: x/finality/keeper/power_dist_change.go
Comment (RafilxTenfen) [reply x/finality/keeper/power_dist_change.go:360]: `cacheFpByBtcPkHex` is never sorted, it is just a local cache to get FPs by their BTC PK, the slice of BTC delegations that is sorted to make sure it is deterministic

File: x/finality/keeper/power_dist_change.go
Comment (gitferry) [reply x/finality/keeper/power_dist_change.go:361]: sorry, typo: `so that unbonding events will processed twice by the incentive module` -> `so that unbonding events will not be processed twice by the incentive module`

I think the current impl is good

Comment (RafilxTenfen) [reply x/finality/keeper/power_dist_change.go:361]: So, maybe the removal of the BtcDelegations will be done in a further PR 
cc @SebastianElvis

File: x/finality/keeper/power_dist_change.go
Comment (RafilxTenfen) [reply x/finality/keeper/power_dist_change.go:286]: Can this be done in another PR after this one

File: x/finality/keeper/power_dist_change.go
Comment (RafilxTenfen) [reply x/finality/keeper/power_dist_change.go:361]: That makes total sense to me, I will make the modifications

File: proto/babylon/incentive/rewards.proto
Comment (SebastianElvis) [reply proto/babylon/incentive/rewards.proto:64]: Sure, let's create an issue for now

File: test/e2e/initialization/node.go
Comment (SebastianElvis) [inline test/e2e/initialization/node.go:358]: probably we can keep it at debug level for debugging purposes

File: x/finality/keeper/power_dist_change.go
Comment (SebastianElvis) [reply x/finality/keeper/power_dist_change.go:286]: Could we do it in this PR or merge this PR to a feature branch? imo replacing `BtcDels` with the "virtual periods" is the main point of this feature so it's better to have it before merging this feature to `main`

File: x/finality/keeper/power_dist_change.go
Comment (SebastianElvis) [reply x/finality/keeper/power_dist_change.go:353]: I see. Probably a better fix is then distinguishing between two types of unbonded events, but this is certainly out of this PR's scope.

Also, why is `newActiveBtcDels` needed given that a BTC del cannot be active twice?

File: x/finality/keeper/power_dist_change.go
Comment (RafilxTenfen) [reply x/finality/keeper/power_dist_change.go:360]: modified at https://github.com/babylonlabs-io/babylon/pull/306/commits/dcee28ef1438a50c6fdc060b4e559993aa3151a3

File: proto/babylon/incentive/rewards.proto
Comment (RafilxTenfen) [reply proto/babylon/incentive/rewards.proto:64]: https://github.com/babylonlabs-io/babylon/issues/357

File: x/incentive/keeper/reward_tracker.go
Comment (RafilxTenfen) [reply x/incentive/keeper/reward_tracker.go:88]: It iterates over all the delegations from one address and acts as if the btc delegation was modified (which was, since BTC rewards are being withdraw) it just doesn't modifies the amount staked

File: x/finality/keeper/power_dist_change.go
Comment (RafilxTenfen) [reply x/finality/keeper/power_dist_change.go:352]: I didn't had in mind that the events were deterministic, my bad 

I can process the reward tracker right at the moment the event are received, I will update tomorrow and remove the `newActiveBtcDels` and `newUnbondedBtcDels`

File: x/finality/keeper/power_dist_change.go
Comment (RafilxTenfen) [reply x/finality/keeper/power_dist_change.go]: Removed `BTCDelDistInfo` structure

File: x/finality/keeper/power_dist_change.go
Comment (RafilxTenfen) [reply x/finality/keeper/power_dist_change.go]: removed

File: x/incentive/keeper/reward_tracker.go
Comment (RafilxTenfen) [reply x/incentive/keeper/reward_tracker.go:279]: I will create the issue for the reference count for pruning 
and another one for removing used `GetBTCStakingGauge` which stores the amount of rewards distributed every block https://github.com/babylonlabs-io/babylon/blob/96f5a1fd34d6b09697bbb7fa7b4f5090fc864d2f/x/incentive/keeper/btc_staking_gauge.go#L17

But once it is used, should not be in the database anymore, if anyone wants to know the amounts from old heights it should index events

File: x/finality/keeper/power_dist_change.go
Comment (RafilxTenfen) [reply x/finality/keeper/power_dist_change.go:354]: I will add another explicity test case for this as well

File: x/finality/types/power_table.go
Comment (RafilxTenfen) [reply x/finality/types/power_table.go]: modified functions to sats amounts at https://github.com/babylonlabs-io/babylon/pull/306/commits/33f1b7f0ee42b5ffa6867cf04f171743db2a7885

Issue to add test https://github.com/babylonlabs-io/babylon/issues/361

File: x/incentive/keeper/reward_tracker.go
Comment (RafilxTenfen) [reply x/incentive/keeper/reward_tracker.go:279]: Reference count https://github.com/babylonlabs-io/babylon/issues/362
`BTCStakingGauge` https://github.com/babylonlabs-io/babylon/issues/363

File: x/finality/keeper/power_dist_change.go
Comment (RafilxTenfen) [reply x/finality/keeper/power_dist_change.go:353]: test added https://github.com/babylonlabs-io/babylon/pull/306/commits/20c1c08f5a4a3014ee3789370a641463f35248f9

File: x/btcstaking/types/btc_delegation.go
Comment (RafilxTenfen) [reply x/btcstaking/types/btc_delegation.go:146]: good catch, done at https://github.com/babylonlabs-io/babylon/pull/306/commits/5ed4b1c43d41ac8f149252806dfcb0c7c54a43e6

**Review by KonradStaniec — APPROVED** [review]
looks good to me 👍 great work pushing it forward :) 

Lets wait for at leas one more approval before merging given size of the pr  :)

File: x/incentive/keeper/reward_tracker_store.go
Comment (RafilxTenfen) [reply x/incentive/keeper/reward_tracker_store.go:1]: modified to use collections in most parts

**Comment by RafilxTenfen** [comment]
@SebastianElvis @gitferry Addressed all the comments :rocket:

File: x/finality/keeper/power_dist_change.go
Comment (gitferry) [inline x/finality/keeper/power_dist_change.go:286]:
```suggestion
		// process all new BTC delegations under this finality provider
```
Aren't them unbonding?

File: x/finality/keeper/power_dist_change.go
Comment (gitferry) [inline x/finality/keeper/power_dist_change.go:290]: here we process unbondedSats earlier than bondedSats, will it make the sats overflow (uin64 cannot be a negative value)? Better have a test case for it

File: x/finality/keeper/power_dist_change.go
Comment (gitferry) [inline x/finality/keeper/power_dist_change.go:336]: should use sats other than dels

File: x/finality/keeper/power_dist_change.go
Comment (gitferry) [inline x/finality/keeper/power_dist_change.go:349]:
```suggestion
		fpUnbondedBTCDels := unbondedSatsByFpBtcPk[fpBTCPKHex]
		for _, unbodedSats := range fpUnbondedBTCDels {
			fpDistInfo.RemoveBondedSats(unbodedSats)
		}
```

**Review by gitferry — APPROVED** [review]
Great work on the improvement! Some minor comments, no blocker:

File: x/finality/keeper/power_dist_change.go
Comment (RafilxTenfen) [reply x/finality/keeper/power_dist_change.go]: modified to sats

File: x/finality/keeper/power_dist_change.go
Comment (RafilxTenfen) [reply x/finality/keeper/power_dist_change.go:302]: modified to handle active first, test should be added either way in another PR :sweat_smile:

File: x/finality/keeper/power_dist_change.go
Comment (RafilxTenfen) [reply x/finality/keeper/power_dist_change.go:286]: good catch

File: proto/babylon/btcstaking/v1/btcstaking.proto
Comment (RafilxTenfen) [reply proto/babylon/btcstaking/v1/btcstaking.proto]: updated

**Review by Lazar955 — APPROVED** [review]
Amazing work!

**Pull request merged**
